bofire/surrogates/mixed_tanimoto_gp.py

```diff
@@ -0,0 +1,369 @@
+from gpytorch.likelihoods.likelihood import Likelihood
+from gpytorch.mlls import ExactMarginalLogLikelihood
+from gpytorch.priors import GammaPrior
+from torch import Tensor
+
+import bofire.kernels.api as kernels
+from bofire.data_models.enum import (
+    CategoricalEncodingEnum,
+    OutputFilteringEnum,
+)
+from bofire.data_models.molfeatures.api import (
+    Fingerprints,
+    FingerprintsFragments,
+    Fragments,
+    MordredDescriptors,
+)
+
+# MolecularEncodingEnum,
+from bofire.data_models.surrogates.api import MixedTanimotoGPSurrogate as DataModel
+
+# from bofire.data_models.kernels.categorical import HammondDistanceKernel
+from bofire.surrogates.botorch import BotorchSurrogate
+from bofire.surrogates.single_task_gp import get_scaler
+from bofire.surrogates.trainable import TrainableSurrogate
+from bofire.utils.torch_tools import tkwargs
+
+
+class MixedTanimotoGP(SingleTaskGP):
+    def __init__(
+        self,
+        train_X: Tensor,
+        train_Y: Tensor,
+        mol_dims: List[int],
+        mol_kernel_factory: Callable[[torch.Size, int, List[int]], Kernel],
+        cat_dims: Optional[List[int]] = None,
+        # cat_kernel_factory: Optional[
+        #    Callable[[torch.Size, int, List[int]], Kernel]
+        # ] = None, --> BoTorch forced to use CategoricalKernel
+        cont_kernel_factory: Optional[
+            Callable[[torch.Size, int, List[int]], Kernel]
+        ] = None,
+        likelihood: Optional[Likelihood] = None,
+        outcome_transform: Optional[OutcomeTransform] = None,
+        input_transform: Optional[InputTransform] = None,
+    ) -> None:
+        if len(mol_dims) == 0:
+            raise ValueError("Must specify molecular dimensions for MixedTanimotoGP")
+
+        if cat_dims is None:
+            cat_dims = []
+        self._ignore_X_dims_scaling_check = cat_dims
+
+        _, aug_batch_shape = self.get_batch_dimensions(train_X=train_X, train_Y=train_Y)
+
+        d = train_X.shape[-1]
+        mol_dims = normalize_indices(indices=mol_dims, d=d)
+        cat_dims = normalize_indices(indices=cat_dims, d=d)
+        ord_dims = sorted(set(range(d)) - set(cat_dims) - set(mol_dims))  # type: ignore
+
+        if cont_kernel_factory is None:
+            cont_kernel_factory = kernels.map_MaternKernel(  # type: ignore
+                data_model=MaternKernel(ard=True, nu=2.5),
+                batch_shape=aug_batch_shape,
+                ard_num_dims=len(ord_dims),
+                active_dims=ord_dims,
+            )
+
+        if likelihood is None:
+            min_noise = 1e-5 if train_X.dtype == torch.float else 1e-6
+            likelihood = GaussianLikelihood(
+                batch_shape=aug_batch_shape,
+                noise_constraint=GreaterThan(
+                    min_noise, transform=None, initial_value=1e-3
+                ),
+                noise_prior=GammaPrior(0.9, 10.0),
+            )
+
+        if len(ord_dims) == 0:
```

Comment:
@simonsung06: Do we really want to allow here for no continuous features? For me it is then not anymore a "MixedGP" but a completely categorical GP, which we also need to implement ...

Reply:
Fine for me!